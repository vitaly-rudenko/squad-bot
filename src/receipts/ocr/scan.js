import { env } from '../../env.js'

/**
 * @param {import('../types').Photo} photo
 * @returns {Promise<number[] | undefined>}
 */
export async function scan(photo) {
  const form = new FormData()
  form.set('file', new Blob([photo.buffer], { type: photo.mimetype }), 'photo.jpg')
  form.set('detectOrientation', 'true')
  form.set('isTable', 'true')
  form.set('scale', 'true')
  form.set('OCREngine', '2')

  const response = await fetch(env.OCR_SPACE_ENDPOINT, {
    method: 'POST',
    headers: {
      Apikey: env.OCR_SPACE_API_KEY,
    },
    body: form,
  })

  /** @type {any} */
  const results = await response.json()

  /** @type {string} */
  const parsedText = results?.ParsedResults?.[0]?.ParsedText
  if (!parsedText) return undefined

  return extractAmounts(parsedText)
}

const AMOUNT_SEARCH_REGEX = /(\b| |^)(([\dззiobs]{1,2} )[\dззiobs,\.]{3,}|[\dзiobs,\.]+)(\b| |$)/g
const AMOUNT_VALIDATION_REGEX = /^\d+(\.\d{2})?$/
const WHITESPACE_LIKE_REGEX = /[ \f\n\r\t\v\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]+/g

/** @param {string} text */
export function extractAmounts(text) {
  text = text.toLowerCase().replaceAll(WHITESPACE_LIKE_REGEX, ' ')

  const matches = text.match(AMOUNT_SEARCH_REGEX)
  if (!matches) return []

  const currentYear = String(new Date().getFullYear())

  return matches
    .map(match => match
      .trim()
      .replaceAll(',', '.')
      .replaceAll('з', '3')
      .replaceAll('з', '3')
      .replaceAll('i', '1')
      .replaceAll('o', '0')
      .replaceAll('s', '5')
      .replaceAll('b', '6'))
    .flatMap(number => number.includes(' ') ? [number.replaceAll(' ', ''), ...number.split(' ')] : [number])
    .filter(number => AMOUNT_VALIDATION_REGEX.test(number) && !number.startsWith('0') && number !== currentYear)
    .sort((a, b) => scoreAmount(b) - scoreAmount(a))
    .map(Number)
    .filter(amount => Number.isFinite(amount) && amount > 0 && amount <= 100_000)
    .map(amount => amountToCoins(amount))
    .filter((value, index, self) => self.indexOf(value) === index)
}

/** @param {string} amount */
function scoreAmount(amount) {
  return amount.includes('.') ? 1 : 0
}

/** @param {number} number */
function amountToCoins(number) {
  return Number((number * 100).toFixed(0))
}
