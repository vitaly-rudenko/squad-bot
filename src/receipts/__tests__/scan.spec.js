import { extractAmounts } from '../scan'
import { receipt1, receipt2, receipt3, receipt4, receipt5, receipt6, receipt7 } from './fixtures'

describe('scan', () => {
  describe('extractAmounts()', () => {
    it('extracts amounts from text', () => {
      const currentYear = String(new Date().getFullYear())

      expect(extractAmounts(`
        Date: 12.03.${currentYear} (March 12th ${currentYear}) at 12:40 (UTC+3)
        Customers: 2

        Service 1 0.00
        Chicken soup 1 2.50
        Beef soup 2 124,46
        Water 3 0.50

        Total: 127.46
        Tip: 20
      `)).toIncludeSameMembers([250, 212446, 12446, 12746, 300, 50, 200, 100, 2000])
    })

    it('handles ambiguous thousands', () => {
      expect(extractAmounts('1 799,00')).toIncludeAllMembers([1799_00, 100, 799_00])
    })

    it('handles wrongly detected characters', () => {
      expect(extractAmounts('Ð·O2I')).toIncludeAllMembers([3021_00])
      expect(extractAmounts('b0s')).toIncludeAllMembers([605_00])
    })

    it('handles ambiguous amounts that might be missing point', () => {
      expect(extractAmounts('79900')).toIncludeAllMembers([799_00, 79900_00])
      expect(extractAmounts('56099')).toIncludeAllMembers([560_99, 56099_00])
    })

    it('parses real receipts', () => {
      expect(extractAmounts(receipt1)).toIncludeAllMembers([386_76, 351_60, 921_36, 1140_48, 3026_73, 3254_95, 3100_00, 3181_68])
      expect(extractAmounts(receipt2)).toIncludeAllMembers([139_00, 229_00, 198_00, 149_00, 554_00, 1418_00])
      expect(extractAmounts(receipt3)).toIncludeAllMembers([364_69, 6_66, 37_79, 140_00, 50_50, 82_90, 53_50])
      expect(extractAmounts(receipt4)).toIncludeAllMembers([48_00, 96_00, 159_00, 93_00, 170_00, 518_00])
      expect(extractAmounts(receipt5)).toIncludeAllMembers([168_00, 25_00, 95_00, 155_00, 275_00, 443_00])
      expect(extractAmounts(receipt6)).toIncludeAllMembers([65_00, 105_00, 180_00, 360_00, 530_00])
      expect(extractAmounts(receipt7)).toIncludeAllMembers([4_50, 78_78, 51_50, 28_32, 60_19, 31_78, 44_98, 32_29])
    })

    it('parses real receipts (snapshots)', () => {
      expect(extractAmounts(receipt1)).toMatchSnapshot()
      expect(extractAmounts(receipt2)).toMatchSnapshot()
      expect(extractAmounts(receipt3)).toMatchSnapshot()
      expect(extractAmounts(receipt4)).toMatchSnapshot()
      expect(extractAmounts(receipt5)).toMatchSnapshot()
      expect(extractAmounts(receipt6)).toMatchSnapshot()
      expect(extractAmounts(receipt7)).toMatchSnapshot()
    })
  })
})