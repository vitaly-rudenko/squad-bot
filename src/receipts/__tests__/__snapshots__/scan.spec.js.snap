// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan extractAmounts() parses real receipts (snapshots) 1`] = `
[
  38676,
  35160,
  92136,
  114048,
  14048,
  302673,
  325495,
  25495,
  310000,
  10000,
  318168,
  18168,
  1800,
  500,
  100,
  300,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 2`] = `
[
  123456,
  100,
  13900,
  22900,
  200,
  19800,
  14900,
  55400,
  141800,
  800,
  300,
  589100,
  1000,
  1500,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 3`] = `
[
  36469,
  666,
  3779,
  14000,
  5050,
  8290,
  5350,
  1000,
  10000,
  13500,
  45000,
  300,
  200,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 4`] = `
[
  155893,
  4800,
  9600,
  15900,
  9300,
  17000,
  51800,
  151800,
  123456,
  3100,
  12500,
  300,
  10000,
  600,
  19000,
  100,
  2100,
  2000,
  1200,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 5`] = `
[
  123456,
  1429,
  816800,
  16800,
  2500,
  9500,
  115500,
  15500,
  27500,
  44300,
  10100,
  100,
  300,
  800,
  200,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 6`] = `
[
  18310,
  6500,
  110500,
  10500,
  218000,
  18000,
  36000,
  53000,
  1831000,
  100,
  300,
  200,
]
`;

exports[`scan extractAmounts() parses real receipts (snapshots) 7`] = `
[
  450,
  7878,
  5150,
  2832,
  6019,
  3178,
  4498,
  3229,
  3700,
  33000,
  33100,
  300,
  1200,
  100,
  200,
  6500,
  4800,
  2800,
  30000,
]
`;
